### PR TITLE
Don't attach multiple event listeners or lifecycle listeners on multiple calls to `onStart`

### PR DESCRIPTION
**Pull request by ramatevish**
Retry logic added prior relies on multiple calls to `onStart`, but this results in duplicate  listeners being added to the underlying mysql client, causing duplicate publishing of events, and duplicate calls in response to lifecycle events. This PR aims to only attached relevant logic once, though we do attach multiple `ConnectionNotifyingLifecycleListener` instances to track the individual connection successes.

~~I'm presently in the process of testing this in sand~~
Tested in sand (reviewed version had lazy-eval problem), booted sand, loaded two sessions, restarted mysql, confirmed from logs that onConnect/onDisconnect works, confirmed invalidator container resumed serving invalidations and that both sessions had expected reactivity.

### CODE REVIEW

**Comment by arvindvjkumar** [comment]
Just as a note, a possibly safer way to solve this is to just recreate the entire consumer on each retry (in `LunaDbBinlogConsumerImpl`)

File: mypipe-api/src/main/scala/mypipe/mysql/AbstractMySQLBinaryLogConsumer.scala [hunk 1]
```diff
@@ -130,41 +144,41 @@ abstract class AbstractMySQLBinaryLogConsumer
...
 
     val startTime = Platform.currentTime
-    while (Platform.currentTime - startTime < connectWaitTime && !connected) Thread.sleep(10)
+    while (Platform.currentTime - startTime < connectWaitTime && !isConnected()) Thread.sleep(10)
 
-    Future.successful(connected)
+    Future.successful(isConnected())
   }
 
   override protected def onStop(): Unit = client.disconnect()
```
Comment (arvindvjkumar) [inline mypipe-api/src/main/scala/mypipe/mysql/AbstractMySQLBinaryLogConsumer.scala:184]: Hmm, do we have ordering guarantees? In other words, is it possible for this to happen:
```
- We call onStop, ie. disconnect client
- We call onStart, ie. setup new client lifecycle listener
- Client updates listeners with connect messages
- Client updates listeners with disconnect message
```

I'm guessing this is likely not possible though. Also if client.disconnect is blocking() this should probably be fine.

**Review by arvindvjkumar — APPROVED** [review]
Overall, I think this is ok, but making the retry code simpler (ie. just recreate the client) might be safer

**Comment by ramatevish** [comment]
> Overall, I think this is ok, but making the retry code simpler (ie. just recreate the client) might be safer

This was our conclusion as well, but we'd like to get a fix out ASAP so we're not under the gun.

File: mypipe-api/src/main/scala/mypipe/mysql/AbstractMySQLBinaryLogConsumer.scala (hunk 1 above)
Comment (ramatevish) [reply mypipe-api/src/main/scala/mypipe/mysql/AbstractMySQLBinaryLogConsumer.scala:184]: I had the same thought, though If that was the case, I believe the current code would be vulnerable to this potential race condition as well. Re-writing the retry logic entirely (potentially one level up), is probably the better option medium term.

**Pull request merged**